File: Python_course/Tasks/Algorithms/Binary_search/Binary_search.py
# Двоичный поиск
def binary_search(arr, key):
    low = 0
    high = len(arr) - 1
    while low <= high:
        mid = (low + high) // 2
        guess = arr[mid]
        if guess == key:
            return mid
        elif guess < key:
            low = mid + 1
        else:
            high = mid - 1
    return None

File: Fundamentals_of_Software_Design/C_language/Binary_search/binary_search.c
#include "binary_search.h"

int binary_search(const int* arr, size_t size, int key) {
    if (arr == NULL || size == 0) {
        return -1;
    }

    size_t left = 0;
    size_t right = size - 1;

    while (left <= right) {
        size_t mid = left + (right - left) / 2;
        
        if (arr[mid] == key) {
            return mid;
        }
        
        if (arr[mid] < key) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    return -1;
}
